### ISSUE TITLE
Versioning: `premajor` broken and doesn't match `--dry-run`

### ISSUE THREAD

**Issue by Jason3S** [lerna-lite/lerna-lite#569]
### Describe the bug

Dry Run and Actual results in different versions:

Command:
```sh
lerna version --conventional-commits --sync-workspace-lock --yes --conventional-prerelease --dry-run premajor
```

Version: `6.31.1 => 7.0.0-alpha.0`

<img width="939" alt="image" src="https://user-images.githubusercontent.com/3740137/232016620-1843c63b-e611-4f40-9781-28a633ef671f.png">

Command:
```sh
lerna version --conventional-commits --sync-workspace-lock --yes --conventional-prerelease premajor
```

Version: `6.31.1 => 6.31.2`

<img width="882" alt="image" src="https://user-images.githubusercontent.com/3740137/232017487-7b76a763-86bc-4969-bfdc-ec6e51bd069d.png">

### Expectation

Expect that the version generated without `--dry-run` matches the dry-run version.

### Reproduction

There seems to be two different issues here:
1. `--dry-run` and non-dry-run do not match.
2. `premajor` and or `--conventional-prerelease` do not seem to be working.

If needed, I'll make a minimal reproduction, but given the obvious issue here, it is already something to be handled by the Unit Tests.

### Lerna config and logs

<details><summary>lerna.json</summary><p>

```json
{
  "npmClient": "pnpm",
  "packages": [
    "./packages/*",
    "./rfc/*",
    "./test-packages/*/*",
    "./integration-tests"
  ],
  "version": "6.31.2",
  "command": {
    "version": {
      "allowBranch": "main",
      "conventionalCommits": true
    }
  }
}
```
</p></details>

### Environment Info

```
System:
    OS: Linux 5.15 Ubuntu 22.04.2 LTS 22.04.2 LTS (Jammy Jellyfish)
    CPU: (2) x64 Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
    Memory: 5.87 GB / 6.78 GB
    Container: Yes
    Shell: 5.1.16 - /bin/bash
  Binaries:
    Node: 16.20.0 - /opt/hostedtoolcache/node/16.20.0/x64/bin/node
    Yarn: 1.22.[19](https://github.com/streetsidesoftware/cspell/actions/runs/4699014859/jobs/8331987269#step:13:20) - /usr/local/bin/yarn
    npm: 8.19.4 - /opt/hostedtoolcache/node/16.[20](https://github.com/streetsidesoftware/cspell/actions/runs/4699014859/jobs/8331987269#step:13:21).0/x64/bin/npm
  Browsers:
    Chrome: 112.0.5615.49
    Chromium: 112.0.5593.0
    Firefox: 112.0
  npmPackages:
    @lerna-lite/cli: ^1.17.0 => 1.17.0
```

### Used Package Manager

pnpm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/lerna-lite/lerna-lite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [docs](https://github.com/lerna-lite/lerna-lite#readme).
- [X] Check that there isn't [already an issue](https://github.com/lerna-lite/lerna-lite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/lerna-lite/lerna-lite/discussions).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by ghiscoding** [comment]
you can't put `premajor` at the end, it won't be recognized... for it to be detected as the `bump` it must be right after `version` or `publish` and in your case it looks like you're padding `premajor` to the `--dry-run` option which obviously won't do anything with this argument

```sh
lerna version premajor --conventional-commits --sync-workspace-lock --yes --conventional-prerelease --dry-run
```

as per the CLI

https://github.com/lerna-lite/lerna-lite/blob/e787a5aaf30c7287382e31b120073b9b38e9c694/packages/cli/src/cli-commands/cli-version-commands.ts#L21-L23

**Comment by ghiscoding** [comment]
and in the case of Lerna-Lite v2.0.0 which I just released last night but pre-released a week ago, I used this command for the pre-release (you can adapt for `lerna version`, it should work the same)

```sh
lerna publish 2.0.0-alpha.0 --dist-tag next --dry-run
```

You can also see this pre-release on the project
https://github.com/lerna-lite/lerna-lite/releases/tag/v2.0.0-alpha.2

**Comment by Jason3S** [comment]
@ghiscoding,

I'm not understanding why it "worked" with `--dry-run` but not without.

**Comment by ghiscoding** [comment]
in your use case, because `premajor` was at the end of your CLI command, it was simply not being used at all. I assume that because it wasn't detected, then conventional commits was taking over and suggesting the next version for you.

I tried with `premajor` at the beginning and it works as expected, so I'll close this issue since it seems to work as expected when used correctly. The result is the same with/without `--dry-run` which is confirming that it works as intended when `premajor` is following `lerna version premajor`

```sh
lerna version premajor --conventional-commits --sync-workspace-lock --conventional-prerelease
```

![image](https://user-images.githubusercontent.com/643976/232168288-5a89b547-5a42-4009-80bc-66326dbf8ebb.png)

Also to provide more info, the CLI is built with yargs and the `bump` is a called a positional argument and must be provided right after the command as per the yargs [docs](https://github.com/yargs/yargs/blob/main/docs/advanced.md#providing-a-command-module) (I highlighted below the important part)
> `command`: string (or array of strings) that executes this command when given on the command line, **first string may contain positional args**

the position argument in our case is `[bump]` from the version CLI below

https://github.com/lerna-lite/lerna-lite/blob/e787a5aaf30c7287382e31b120073b9b38e9c694/packages/cli/src/cli-commands/cli-version-commands.ts#L21-L22

**Comment by Jason3S** [comment]
@ghiscoding,

Thank you for the detailed response to: 

> 2. `premajor` and or `--conventional-prerelease` do not seem to be working.

In my case, it wasn't a feature I used for a long time. `lerna version` is part of a deploy workflow script where the conventional commits normally does the right thing. But it wasn't picking up the breaking change, so I had to add the bump. This is one of those things that "used to work", but since I hadn't needed to add the bump for so long, I wasn't aware when it stopped working. So, I was using `--dry-run` to experiment with the arguments and options until I thought it was working.

> 1. `--dry-run` and non-dry-run do not match.

I was more surprised that `--dry-run` changed the versioning behavior. I was relying on `--dry-run` to show me the result I could expect. That part, I believe is still broken.

Please see the original images above:

Dry Run reported a version change: `6.31.1 => 7.0.0-alpha.0`
But the actual run had a different result: `6.31.1 => 6.31.2`

I'm very happy you added `--dry-run`, it made it possible to test out workflows. But, I need to trust that `--dry-run` is giving me the right information.

**Comment by ghiscoding** [comment]
but like I said you were passing `premajor` to the `--dry-run` flag which is expecting a boolean but you're giving it a string, so I'm not surprised it was acting weird. The dry-run is working totally fine if you do put the bump at the correct location. Before closing the issue, I did try it

> This is one of those things that "used to work", but since I hadn't needed to add the bump for so long, I wasn't aware when it stopped working.

it didn't stop working, you simply didn't use it the way it expects it. The main reason it wasn't working in your case is like I said is that the bump must follow the command and it has always been this way. I've seen someone opening a similar issue on the original Lerna and it was closed by Lerna's maintainer with the same answer but I just can't find the issue to reference it.... so again it always worked this way and it's how the yargs library is working and it's not broken at all. 

The dry-run is working fine, I don't need to investigate it, just use it like I said and it's working totally fine, there are multiple unit tests covering version bump

**Comment by Jason3S** [comment]
@ghiscoding,

Thank you for your thoughtful reply. I did specify the arguments in the wrong order. Please note: I made a common mistake that others might make. I wrongly assumed it matched the common CLI pattern: `app command [options] <args>`.

Even so, whether or not I specified the arguments in the wrong order is immaterial. I have already changed the order in my scripts. This issue is and was always about **trust**. How would you feel if a `remove` app had a `--dry-run` option that shows which files would be deleted, then it deleted more files? Would you trust that app? In this case `lerna-lite` told me it would generate the version I wanted, but when I removed `--dry-run` it did something different.

I realize you made this fork for the community. Thank you.

If you are open to it, I'll look into this issue and make a PR that addresses:
1. Why did the behavior change when `--dry-run` was removed.
2. (optionally) Improve the argument checking and report when things are not expected. 

Kind regards,
Jason

**Comment by ghiscoding** [comment]
Sure I'm totally open for PR, I understand your point but yes I spent a considerable amount of time creating this fork and I want to take a bit of break. Because I am not paid at all for working on this and the last v2.0 release was quite a lot of work to switch to ESM and Vitest, it took me well over a month and a lot of night to go to bed at 2AM, so yes I want to take a little break of this and this issue was less of a priority in my mind. It would be nice to get more help from the community indeed :)

Thanks for the feedback

**Comment by ghiscoding** [comment]
> 1. `--dry-run` changes the behavior based upon its location on the command line. There are a couple of ways to handle this. A simple one would be to remove the `boolean` type and give an error if anything was assigned to it. Another would be to make sure it was always the last argument.

I don't think you can remove the fact that it's a boolean, when you pass `--dry-run` then yargs will automatically parse it as a boolean and the boolean is also required to be provided when used in the `lerna.json` config, so the answer is No we can't remove its boolean type. As for making it the last argument, I don't think there's any way to do with yargs or any other lib for that matter. The only thing that might be possible is to check if the argument provided is a boolean or something else, but I'm not sure if it's possible with yargs or if it will automatically parse it to true when a string is provided, I taught the CLI would throw an error when using incorrect type but I could be wrong

https://github.com/lerna-lite/lerna-lite/blob/f083ee6ff9245cae767546e0d0fcce03238f59db/packages/cli/src/cli-commands/cli-version-commands.ts#L81-L85

> 2\. `--conventional-prerelease` is not checking to see if any packages match the value. This seems like a needed check. Misspelling a package name means it can be released as `latest` when it was expected to be `alpha`.

That came from Lerna implementation and I never used it, but if I look at the documentation it should be possible to not provide any argument because they made it as a loose type and they did mention it could be a boolean (that is when no argument is provided like you did, so it would parse it as a boolean). 

https://github.com/lerna-lite/lerna-lite/blob/f083ee6ff9245cae767546e0d0fcce03238f59db/packages/cli/src/cli-commands/cli-version-commands.ts#L50-L53

There is a check on the `bump` option, but since you didn't provide it at the correct location it didn't get validated and simply ignored

https://github.com/lerna-lite/lerna-lite/blob/f083ee6ff9245cae767546e0d0fcce03238f59db/packages/cli/src/cli-commands/cli-version-commands.ts#L9-L18

**Comment by ghiscoding** [comment]
Now released under [v2.1.0](https://github.com/lerna-lite/lerna-lite/releases/tag/v2.1.0), thanks a lot for the contribution
